streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala

```diff
@@ -350,6 +349,18 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
     )
   }
 
+  /** TODO: Add scala docs */
+  def trackStateByKey[S: ClassTag, T: ClassTag](
+    spec: TrackStateSpec[K, V, S, T]): EmittedRecordsDStream[K, V, S, T] = {
+    new EmittedRecordsDStreamImpl[K, V, S, T](
+      new TrackStateDStream[K, V, S, T](
+        self,
+        spec.asInstanceOf[TrackStateSpecImpl[K, V, S, T]]
```

Comment:
Seems like this is what we did with DataFrameReader which follows a similar pattern:

https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L47